### PR TITLE
Add API for monitoring built Histograms and improve reproducibility test [nocheck]

### PR DESCRIPTION
**Pull request by michalkurka**

### CODE REVIEW

**Review by maurever — COMMENTED** [review]
Looks good to me. Just one question and one small suggestion. Thanks, @michalkurka.

File: h2o-algos/src/main/java/hex/tree/SharedTree.java [hunk 1]
```diff
@@ -1223,13 +1242,15 @@ public boolean equals(Object o) {
...
-      return _keep_orig_histo_precision == that._keep_orig_histo_precision;
+      if (_keep_orig_histo_precision != that._keep_orig_histo_precision) return false;
+      return _histo_monitor_class != null ? _histo_monitor_class.equals(that._histo_monitor_class) : that._histo_monitor_class == null;
     }
 
     @Override
     public int hashCode() {
       int result = (_reproducible_histos ? 1 : 0);
       result = 31 * result + (_keep_orig_histo_precision ? 1 : 0);
+      result = 31 * result + (_histo_monitor_class != null ? _histo_monitor_class.hashCode() : 0);
```
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/tree/SharedTree.java:1253]: I don't think it does - it takes result from line 1252 into account. The hashCode impl is auto-generated by IntelliJ - it looks fine to me

File: h2o-algos/src/test/java/hex/tree/gbm/GBMTest.java [hunk 2]
```diff
@@ -4731,8 +4729,76 @@ public void testStrictReproducibility() {
...
+    } finally {
+      HistogramCollectingMonitor._histos.clear();
+    }
+  }
+
+  static void compareModelHistos(List<HistogramCollectingMonitor> histos, int histosPerModel) {
+    assertTrue(histosPerModel > 0);
+    for (int i = 0; i < histos.size() - histosPerModel; i++) {
+      HistogramCollectingMonitor histo = histos.get(i);
+      HistogramCollectingMonitor histo_comp = histos.get(i + histosPerModel);
```
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-algos/src/test/java/hex/tree/gbm/GBMTest.java:4753]: will do - thanks!

File: h2o-algos/src/main/java/hex/tree/SharedTree.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (maurever) [reply h2o-algos/src/main/java/hex/tree/SharedTree.java:1253]: Yes, I see! Sorry, my mistake.

**Pull request merged**